docs/optimizing-javascript-loading.md

```diff
@@ -0,0 +1,222 @@
+
+```tsx title="VeryExpensive.tsx"
+import {Component} from 'react';
+import {Text} from 'react-native';
+// ... import some very expensive modules
+
+export default function VeryExpensive() {
+  // ... lots and lots of rendering logic
+  return <Text>Very Expensive Component</Text>;
+}
+```
+
+```tsx title="Optimized.tsx"
+import {useCallback, useState} from 'react';
+import {TouchableOpacity, View, Text} from 'react-native';
+// Usually we would write a static import:
+// import VeryExpensive from './VeryExpensive';
+
+let VeryExpensive = null;
+
+export default function Optimize() {
+  const [needsExpensive, setNeedsExpensive] = useState(false);
+  const didPress = useCallback(() => {
+    if (VeryExpensive == null) {
+      VeryExpensive = require('./VeryExpensive').default;
+    }
+
+    setNeedsExpensive(true);
+  }, []);
+
+  return (
+    <View style={{marginTop: 20}}>
+      <TouchableOpacity onPress={this.didPress}>
+        <Text>Load</Text>
+      </TouchableOpacity>
+      {this.state.needsExpensive ? <VeryExpensive /> : null}
+    </View>
+  );
+}
+```
+
+## Advanced: Automatically inline `require` calls
+
+If you use the React Native CLI to build your app, `require` calls (but not `import`s) will automatically be inlined for you, both in your code and inside any third-party packages (`node_modules`) you use.
+
+```tsx
+import {useCallback, useState} from 'react';
+import {TouchableOpacity, View, Text} from 'react-native';
+
+// This top-level require call will be evaluated lazily as part of the component below.
+const VeryExpensive = require('./VeryExpensive').default;
+
+export default function Optimize() {
+  const [needsExpensive, setNeedsExpensive] = useState(false);
+  const didPress = useCallback(() => {
+    setNeedsExpensive(true);
+  }, []);
+
+  return (
+    <View style={{marginTop: 20}}>
+      <TouchableOpacity onPress={this.didPress}>
+        <Text>Load</Text>
+      </TouchableOpacity>
+      {this.state.needsExpensive ? <VeryExpensive /> : null}
+    </View>
+  );
+}
+```
+
+:::info
+Some React Native frameworks disable this behavior. In particular, in Expo projects, `require` calls are not inlined by default. You can enable this optimization by editing your project's Metro config and setting `inlineRequires: true` in [`getTransformOptions`](https://metrobundler.dev/docs/configuration#gettransformoptions).
+:::
+
+### Pitfalls of inline `require`s
+
+Inlining `require` calls changes the order in which modules are evaluated, and can even cause some modules to _never_ be evaluated. This is usually safe to do automatically, because JavaScript modules are often written to be side-effect-free.
+
+If one of your modules does have side effects - for example, if it initializes some logging mechanism, or patches a global API used by the rest of your code - then you might see unexpected behavior or even crashes. In those cases, you may want to exclude certain modules from this optimization, or disable it entirely.
```

Comment:
Doesn't the transform also skip inlining modules that aren't stored in a variable?

```
import {useCallback, useState} from 'react';
import {TouchableOpacity, View, Text} from 'react-native';
require('ModuleWithSideEffects')
```

Won't that require never be inlined?